internals/overlord/logstate/puller.go

```diff
@@ -0,0 +1,164 @@
+// main control loop.
+type logPuller struct {
+	iterator servicelog.Iterator
+	entryCh  chan<- servicelog.Entry
+
+	tomb tomb.Tomb
+}
+
+// loop pulls logs off the iterator and sends them on the entryCh.
+// The loop will terminate:
+//   - if the puller's context is cancelled
+//   - once the ringbuffer is closed and the iterator finishes reading all
+//     remaining logs.
+func (p *logPuller) loop() error {
+	defer func() { _ = p.iterator.Close() }()
+
+	parser := servicelog.NewParser(p.iterator, parserSize)
+	for p.iterator.Next(p.tomb.Dying()) {
+		for parser.Next() {
+			if err := parser.Err(); err != nil {
+				return err
+			}
+
+			select {
+			case p.entryCh <- parser.Entry():
+			case <-p.tomb.Dying():
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+// pullerGroup represents a group of logPullers, and provides methods for a
+// gatherer to manage logPullers (dynamically add/remove, kill all, wait for
+// all to finish).
+type pullerGroup struct {
+	targetName string
+
+	// Currently active logPullers, indexed by service name
+	pullers map[string]*logPuller
+	// Mutex for pullers map
+	mu sync.RWMutex
+
+	tomb tomb.Tomb
+}
+
+func newPullerGroup(targetName string) *pullerGroup {
+	pg := &pullerGroup{
+		targetName: targetName,
+		pullers:    map[string]*logPuller{},
+	}
+	pg.tomb.Go(func() error {
+		<-pg.tomb.Dying()
+		return nil
+	})
+	return pg
+}
+
+func (pg *pullerGroup) Add(serviceName string, buffer *servicelog.RingBuffer, entryCh chan<- servicelog.Entry) {
+	pg.mu.Lock()
+	defer pg.mu.Unlock()
+
+	// There shouldn't already be a puller for this service, but if there is,
+	// shut it down first and wait for it to die.
+	pg.remove(serviceName)
+
+	lp := &logPuller{
+		iterator: buffer.TailIterator(),
+		entryCh:  entryCh,
+	}
+	lp.tomb.Go(lp.loop)
+	pg.tomb.Go(lp.tomb.Wait)
+
+	pg.pullers[serviceName] = lp
+}
+
+// List returns a list of all service names for which we have a currently
+// active puller.
+func (pg *pullerGroup) List() []string {
+	pg.mu.RLock()
+	defer pg.mu.RUnlock()
+
+	var svcs []string
+	for svc := range pg.pullers {
+		svcs = append(svcs, svc)
+	}
+	return svcs
+}
+
+func (pg *pullerGroup) Remove(serviceName string) {
+	pg.mu.Lock()
+	defer pg.mu.Unlock()
+	pg.remove(serviceName)
+}
+
+func (pg *pullerGroup) remove(serviceName string) {
+	puller, pullerExists := pg.pullers[serviceName]
+	if !pullerExists {
+		return
+	}
+
+	puller.tomb.Kill(nil)
+	delete(pg.pullers, serviceName)
+
+	err := puller.tomb.Wait()
+	if err != nil {
+		logger.Noticef("Error from log puller: %v", err)
+	}
+}
+
+func (pg *pullerGroup) KillAll() {
+	pg.mu.RLock()
+	defer pg.mu.RUnlock()
+
+	for _, puller := range pg.pullers {
+		puller.tomb.Kill(nil)
+	}
+	pg.tomb.Kill(nil)
+}
+
+// Done returns a channel which can be waited on until all pullers have finished.
+func (pg *pullerGroup) Done() <-chan struct{} {
+	return pg.tomb.Dead()
+}
+
+func (pg *pullerGroup) Contains(serviceName string) bool {
```

Comment:
Can the only-used-in-test methods be moved to an `export_test.go` file to make it clear they shouldn't be used in production code?

Reply:
The testing here deals a lot with the internals, so it's "white-box" testing done inside the same package. i.e. there is no `logstate_test` package, you'll notice all the tests are inside the `logstate` package. So I don't think `export_test.go` makes sense here.

We could think about moving all the tests to a `logstate_test` package, but I feel like this would be a lot of extra work exporting otherwise-non-exported things, without a clear benefit in my mind.

Another point - why *shouldn't* `Contains` or `Len` be used in production code? Currently, it happens that they are not, but I don't really see an issue if we did need to use these in the future.

Reply:
> why shouldn't Contains or Len be used in production code?

`Contains` and `Len` look a bit dubious, because they're inherently racey. The locking implies they're concurrency-safe, which they are at a low level (no data races), but not at a high level: as soon as you get the length, it could change. Kinda like how `len(channel)` in Go is normally incorrect: as soon as you have the number of buffered elements, it could be sent/received to on another goroutine and your length could be wrong. That's what it stood out to me.

But I don't have strong feelings here, and what you say regarding white-box testing makes sense.

Reply:
ok, you've made a fair case as to why these shouldn't be used in production code. I'll lowercase both methods and add a note saying they are used for testing.